### PR TITLE
Live preview: Researcher

### PR DESCRIPTION
**Pull request by rgarner**
# [Live preview: Researcher](https://trello.com/c/T2MeDlYY/255-13-live-preview-researcher)

> As a researcher
> I want to see how I am affecting the preview as I edit live values
> So that my changes are obvious and I don't have to skip between steps

Using [react-rails](https://github.com/reactjs/react-rails) and [Jest](https://facebook.github.io/jest/), provide a tested `ResearcherPreviews` component which is live on the researcher page, composed of two separate components that correspond to sections on the final preview. Establish an `Output` component to represent the dynamic parts of the content.

On the final preview page, in order to render links generated from the Rails routing, pass an `editLinks` property to the `ResearcherPreviews` component. The `Output` component will use this along with an `attr` name to intelligently render a link to the step for that attribute.

image

### CODE REVIEW

File: app/javascript/packs/hello_react.jsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Run this example by adding <%= javascript_pack_tag 'hello_react' %> to the head of your layout file,
```
Comment (joelcarr) [inline app/javascript/packs/hello_react.jsx]: We need to delete this file.

Comment (joelcarr) [inline app/javascript/packs/hello_react.jsx]: We need to delete this file.

File: app/assets/stylesheets/app/ResearcherPreviews.css [hunk 2]
```diff
@@ -0,0 +1,26 @@
+.ResearcherPreviews {
```
Comment (joelcarr) [inline app/assets/stylesheets/app/ResearcherPreviews.css]: We should probably make this a `scss` file instead along with the rest

**Comment by joelcarr** [comment]
Sorry, that isn't a final approval. I was just approving the changes we had made for my comments.

File: app/assets/stylesheets/barnardos/_reactive-container.scss [hunk 3]
```diff
@@ -0,0 +1,21 @@
+.reactive-container {
+  display: flex;
+
+  @media (min-width: $tablet) {
```
Comment (jeddy3) [inline app/assets/stylesheets/barnardos/_reactive-container.scss]: Is this nested media query redundant?

File: .babelrc [hunk 4]
```diff
@@ -1,11 +1,22 @@
...
-      "targets": {
-        "browsers": ["> 1%", "last 2 versions", "not ie <= 8"]
+    [
+      "env",
+      {
+        "targets": {
+          "browsers": [
+            "> 1%",
+            "last 2 versions",
+            "not ie <= 8"
```
Comment (joelcarr) [inline .babelrc]: We should change this to not worry about IE 9 and 10.

File: app/assets/stylesheets/barnardos/_reactive-preview.scss [hunk 5]
```diff
@@ -0,0 +1,26 @@
+.reactive-preview {
+  position: relative;
+}
+
+.reactive-preview__section {
+  background: whitesmoke;
+  font: 16px/24px;
```
Comment (jeddy3) [inline app/assets/stylesheets/barnardos/_reactive-preview.scss]: Does this shorthand need a font family?

File: app/javascript/components/ResearcherPreviews/index.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    return (
+      <div>
+        <WhoIsDoingTheResearch {...this.state} />
+        <WhereCanIFindOutMore {...this.state} />
+      </div>
+    )
+  }
+}
+
+ResearcherPreviews.propTypes = {
```
Comment (jeddy3) [inline app/javascript/components/ResearcherPreviews/index.js]: 👍

File: .babelrc (hunk 4 above)
Comment (rgarner) [reply .babelrc]: We should, just not here.

File: .babelrc [hunk 7]
```diff
@@ -1,11 +1,22 @@
...
+          ]
+        }
       }
-    }],
-    "stage-2"
+    ],
+    "stage-2",
+    "es2015",
```
Comment (jeddy3) [inline .babelrc]: I think the `es2015` preset is deprecated in favour of the `env` one. They do the same thing and so I think you can safely remove the `es2015` one.

Comment (rgarner) [reply .babelrc]: Good spot. I'll have a look at this in the morning.

File: app/assets/stylesheets/barnardos/_reactive-preview.scss [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  padding: 8px 16px;
+}
+
+.reactive-preview__section:not(:first-child) {
+  margin-top: 8px;
+}
+
+.reactive-preview__highlight {
+  background: moccasin;
+  word-break: break-word;
```
Comment (rgarner) [inline app/assets/stylesheets/barnardos/_reactive-preview.scss]: @joelcarr RubyMine is telling me `break-word` is not a valid value for this. Is RubyMine right?

File: app/helpers/research_sessions_helper.rb [hunk 9]
```diff
@@ -1,4 +1,18 @@
 module ResearchSessionsHelper
+  def component_params(*attrs, final_preview: false)
+    attrs.each_with_object(
+      editLinks: {}, finalPreview: final_preview
+    ) do |attr, hash|
+      hash[attr] = @research_session.send(attr)
```
Comment (rgarner) [inline app/helpers/research_sessions_helper.rb]: `hash` is not a good name, @rgarner

Comment (rgarner) [reply app/helpers/research_sessions_helper.rb]: Suggest `question_paths`?

File: app/assets/stylesheets/barnardos/_reactive-preview.scss (hunk 5 above)
Comment (joelcarr) [reply app/assets/stylesheets/barnardos/_reactive-preview.scss]: Aye, and it needs a `font-style`.

File: app/assets/stylesheets/barnardos/_reactive-preview.scss (hunk 8 above)
Comment (joelcarr) [reply app/assets/stylesheets/barnardos/_reactive-preview.scss]: Here is a list of options for `word-break` available in Chrome: 
image

File: .babelrc (hunk 4 above)
Comment (rgarner) [reply .babelrc]: https://trello.com/c/RGm3Ssw0/282-update-browser-support-in-babel

**Pull request merged**